In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

In [2]:
# Reading Data
data = pd.read_csv('covid.csv')
data.head()

,iso_code,continent,location,date,total_cases,new_cases,total_deaths,new_deaths,total_vaccinations,people_vaccinated,people_fully_vaccinated,total_boosters,new_vaccinations
0,AFG,Asia,Afghanistan,24/02/20,5.0,5.0,NaN,NaN,NaN,NaN,NaN,NaN,NaN
1,AFG,Asia,Afghanistan,25/02/20,5.0,0.0,NaN,NaN,NaN,NaN,NaN,NaN,NaN
2,AFG,Asia,Afghanistan,26/02/20,5.0,0.0,NaN,NaN,NaN,NaN,NaN,NaN,NaN
3,AFG,Asia,Afghanistan,27/02/20,5.0,0.0,NaN,NaN,NaN,NaN,NaN,NaN,NaN
4,AFG,Asia,Afghanistan,28/02/20,5.0,0.0,NaN,NaN,NaN,NaN,NaN,NaN,NaN


### Exploring Data

In [3]:
data.shape

(183583, 13)

In [4]:
data.columns

Index(['iso_code', 'continent', 'location', 'date', 'total_cases', 'new_cases',
       'total_deaths', 'new_deaths', 'total_vaccinations', 'people_vaccinated',
       'people_fully_vaccinated', 'total_boosters', 'new_vaccinations'],
      dtype='object')

In [5]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 183583 entries, 0 to 183582
Data columns (total 13 columns):
 #   Column                   Non-Null Count   Dtype  
---  ------                   --------------   -----  
 0   iso_code                 183583 non-null  object 
 1   continent                172860 non-null  object 
 2   location                 183583 non-null  object 
 3   date                     183583 non-null  object 
 4   total_cases              176745 non-null  float64
 5   new_cases                176532 non-null  float64
 6   total_deaths             158467 non-null  float64
 7   new_deaths               158464 non-null  float64
 8   total_vaccinations       50011 non-null   float64
 9   people_vaccinated        47564 non-null   float64
 10  people_fully_vaccinated  45066 non-null   float64
 11  total_boosters           22431 non-null   float64
 12  new_vaccinations         41060 non-null   float64
dtypes: float64(9), object(4)
memory usage: 18.2+ MB


In [6]:
data.describe()

,total_cases,new_cases,total_deaths,new_deaths,total_vaccinations,people_vaccinated,people_fully_vaccinated,total_boosters,new_vaccinations
count,1.767450e+05,1.765320e+05,1.584670e+05,158464.000000,5.001100e+04,4.756400e+04,4.506600e+04,2.243100e+04,4.106000e+04
mean,3.036670e+06,1.242833e+04,6.242540e+04,163.242036,2.074084e+08,1.026395e+08,8.387583e+07,3.111973e+07,1.150450e+06
std,1.898075e+07,8.829878e+04,3.260393e+05,805.517132,9.091703e+08,4.487351e+08,3.823615e+08,1.422240e+08,4.248478e+06
min,1.000000e+00,0.000000e+00,1.000000e+00,0.000000,0.000000e+00,0.000000e+00,1.000000e+00,1.000000e+00,0.000000e+00
25%,2.459000e+03,0.000000e+00,9.000000e+01,0.000000,7.441225e+05,4.490690e+05,3.449382e+05,1.036050e+04,5.946750e+03
50%,3.261200e+04,7.500000e+01,8.610000e+02,1.000000,5.929734e+06,3.630465e+06,2.874090e+06,8.269670e+05,4.084250e+04
75%,3.539360e+05,1.059000e+03,8.137000e+03,18.000000,3.699011e+07,2.097199e+07,1.727424e+07,6.528614e+06,2.794522e+05
max,5.149754e+08,4.089131e+06,6.240940e+06,18144.000000,1.162462e+10,5.151278e+09,4.668004e+09,1.877161e+09,5.450643e+07


In [7]:
# Droping 'iso_code' column because it is useless for us
data = data.drop(['iso_code'], axis=1)
data.head()

,continent,location,date,total_cases,new_cases,total_deaths,new_deaths,total_vaccinations,people_vaccinated,people_fully_vaccinated,total_boosters,new_vaccinations
0,Asia,Afghanistan,24/02/20,5.0,5.0,NaN,NaN,NaN,NaN,NaN,NaN,NaN
1,Asia,Afghanistan,25/02/20,5.0,0.0,NaN,NaN,NaN,NaN,NaN,NaN,NaN
2,Asia,Afghanistan,26/02/20,5.0,0.0,NaN,NaN,NaN,NaN,NaN,NaN,NaN
3,Asia,Afghanistan,27/02/20,5.0,0.0,NaN,NaN,NaN,NaN,NaN,NaN,NaN
4,Asia,Afghanistan,28/02/20,5.0,0.0,NaN,NaN,NaN,NaN,NaN,NaN,NaN


In [8]:
data.isnull().sum()

continent                   10723
location                        0
date                            0
total_cases                  6838
new_cases                    7051
total_deaths                25116
new_deaths                  25119
total_vaccinations         133572
people_vaccinated          136019
people_fully_vaccinated    138517
total_boosters             161152
new_vaccinations           142523
dtype: int64

In [9]:
data.keys()[3:]

Index(['total_cases', 'new_cases', 'total_deaths', 'new_deaths',
       'total_vaccinations', 'people_vaccinated', 'people_fully_vaccinated',
       'total_boosters', 'new_vaccinations'],
      dtype='object')

### Replacing empty cells from 'total_cases' to 'new_vaccinations' with 0

In [10]:
for key in data.keys()[3:]:
    print(key)
    data[key].fillna(0, inplace=True)

total_cases
new_cases
total_deaths
new_deaths
total_vaccinations
people_vaccinated
people_fully_vaccinated
total_boosters
new_vaccinations


In [11]:
data.isnull().sum()

continent                  10723
location                       0
date                           0
total_cases                    0
new_cases                      0
total_deaths                   0
new_deaths                     0
total_vaccinations             0
people_vaccinated              0
people_fully_vaccinated        0
total_boosters                 0
new_vaccinations               0
dtype: int64

### Now it's time to clean up 'continent' column

In [12]:
# Remaining enteries
data['continent'].count() - data['continent'].isnull().sum()

162137

In [13]:
data[data['continent'].isnull()]['location'].unique()

array(['Africa', 'Asia', 'Europe', 'European Union', 'High income',
       'International', 'Low income', 'Lower middle income',
       'North America', 'Oceania', 'South America', 'Upper middle income',
       'World'], dtype=object)

In [14]:
data['continent'].unique()

array(['Asia', nan, 'Europe', 'Africa', 'North America', 'South America',
       'Oceania'], dtype=object)

In [15]:
# Droping extra columns
for loc in data[data['continent'].isnull()]['location'].unique():
    if loc not in data['continent'].unique() and loc != 'World':
        data.drop(data[data['location'] == loc].index, inplace=True)

In [16]:
data[data['continent'].isnull()]['location'].unique()

array(['Africa', 'Asia', 'Europe', 'North America', 'Oceania',
       'South America', 'World'], dtype=object)

In [17]:
data['continent'].isnull().sum()

5774

### Now seprating Data

In [18]:
continent_world_data = data[data['location'] in 
        data[data['continent'].isnull()]['location'].unique()]
continent_world_data.head()

<ipython-input-18-fcb25a3a8516>:1: DeprecationWarning: elementwise comparison failed; this will raise an error in the future.
  continent_world_data = data[data['location'] in


ValueError: Lengths must match to compare

### Lengths are different

In [19]:
# checkhing lengths
for loc in data[data['continent'].isnull()]['location'].unique():
    length = data[data['location'] == loc]['location'].count()
    print(loc, length, sep='  => ')

Africa  => 811
Asia  => 833
Europe  => 832
North America  => 833
Oceania  => 830
South America  => 802
World  => 833


Asia, North America and World have same lengths

### Seprating All Continents

In [20]:
world_data = data[data['location'] == 'World']
world_data.head()

,continent,location,date,total_cases,new_cases,total_deaths,new_deaths,total_vaccinations,people_vaccinated,people_fully_vaccinated,total_boosters,new_vaccinations
180444,NaN,World,22/01/20,557.0,0.0,17.0,0.0,0.0,0.0,0.0,0.0,0.0
180445,NaN,World,23/01/20,657.0,100.0,18.0,1.0,0.0,0.0,0.0,0.0,0.0
180446,NaN,World,24/01/20,944.0,287.0,26.0,8.0,0.0,0.0,0.0,0.0,0.0
180447,NaN,World,25/01/20,1437.0,493.0,42.0,16.0,0.0,0.0,0.0,0.0,0.0
180448,NaN,World,26/01/20,2120.0,683.0,56.0,14.0,0.0,0.0,0.0,0.0,0.0


In [21]:
asia_data = data[data['location'] == 'Asia']
asia_data.head()

,continent,location,date,total_cases,new_cases,total_deaths,new_deaths,total_vaccinations,people_vaccinated,people_fully_vaccinated,total_boosters,new_vaccinations
8783,NaN,Asia,22/01/20,556.0,0.0,17.0,0.0,0.0,0.0,0.0,0.0,0.0
8784,NaN,Asia,23/01/20,654.0,98.0,18.0,1.0,0.0,0.0,0.0,0.0,0.0
8785,NaN,Asia,24/01/20,937.0,283.0,26.0,8.0,0.0,0.0,0.0,0.0,0.0
8786,NaN,Asia,25/01/20,1429.0,492.0,42.0,16.0,0.0,0.0,0.0,0.0,0.0
8787,NaN,Asia,26/01/20,2105.0,676.0,56.0,14.0,0.0,0.0,0.0,0.0,0.0


In [22]:
north_america_data = data[data['location'] == 'North America']
north_america_data.head()

,continent,location,date,total_cases,new_cases,total_deaths,new_deaths,total_vaccinations,people_vaccinated,people_fully_vaccinated,total_boosters,new_vaccinations
121536,NaN,North America,22/01/20,1.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
121537,NaN,North America,23/01/20,3.0,2.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
121538,NaN,North America,24/01/20,5.0,2.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
121539,NaN,North America,25/01/20,5.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
121540,NaN,North America,26/01/20,8.0,3.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0


Data of World, Asia and Nort America is started from date **22/01/22**

In [23]:
africa_data = data[data['location'] == 'Africa']
africa_data.head()

,continent,location,date,total_cases,new_cases,total_deaths,new_deaths,total_vaccinations,people_vaccinated,people_fully_vaccinated,total_boosters,new_vaccinations
800,NaN,Africa,13/02/20,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
801,NaN,Africa,14/02/20,1.0,1.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
802,NaN,Africa,15/02/20,1.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
803,NaN,Africa,16/02/20,1.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
804,NaN,Africa,17/02/20,1.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0


Data of Africa is started from date **13/02/20** 

In [24]:
europe_data = data[data['location'] == 'Europe']
europe_data.head()

,continent,location,date,total_cases,new_cases,total_deaths,new_deaths,total_vaccinations,people_vaccinated,people_fully_vaccinated,total_boosters,new_vaccinations
53334,NaN,Europe,23/01/20,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
53335,NaN,Europe,24/01/20,2.0,2.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
53336,NaN,Europe,25/01/20,3.0,1.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
53337,NaN,Europe,26/01/20,3.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
53338,NaN,Europe,27/01/20,4.0,1.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0


Data of Europe is started from date **23/01/20**

In [25]:
oceania_data = data[data['location'] == 'Oceania']
oceania_data.head()

,continent,location,date,total_cases,new_cases,total_deaths,new_deaths,total_vaccinations,people_vaccinated,people_fully_vaccinated,total_boosters,new_vaccinations
124893,NaN,Oceania,25/01/20,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
124894,NaN,Oceania,26/01/20,4.0,4.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
124895,NaN,Oceania,27/01/20,5.0,1.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
124896,NaN,Oceania,28/01/20,5.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
124897,NaN,Oceania,29/01/20,6.0,1.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0


Data of Oceania is started from date **25/01/20**

In [26]:
south_america_data = data[data['location'] == 'South America']
south_america_data.head()

,continent,location,date,total_cases,new_cases,total_deaths,new_deaths,total_vaccinations,people_vaccinated,people_fully_vaccinated,total_boosters,new_vaccinations
152629,NaN,South America,22/02/20,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
152630,NaN,South America,23/02/20,2.0,2.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
152631,NaN,South America,24/02/20,2.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
152632,NaN,South America,25/02/20,2.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
152633,NaN,South America,26/02/20,3.0,1.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0


Data of South America is started from date **22/02/20** 

 **Now we will drop rows to get minimum 802 rows**

We will drop rows from World, Asia, North America, Africa, Europe and Oceania
to get 802 rows with **22/02/20** first row date.

In [27]:
world_data = world_data.iloc[31:]
world_data.head()

,continent,location,date,total_cases,new_cases,total_deaths,new_deaths,total_vaccinations,people_vaccinated,people_fully_vaccinated,total_boosters,new_vaccinations
180475,NaN,World,22/02/20,78608.0,1762.0,2459.0,207.0,0.0,0.0,0.0,0.0,0.0
180476,NaN,World,23/02/20,78990.0,382.0,2470.0,11.0,0.0,0.0,0.0,0.0,0.0
180477,NaN,World,24/02/20,79558.0,568.0,2630.0,160.0,0.0,0.0,0.0,0.0,0.0
180478,NaN,World,25/02/20,80412.0,854.0,2710.0,80.0,0.0,0.0,0.0,0.0,0.0
180479,NaN,World,26/02/20,81385.0,973.0,2771.0,61.0,0.0,0.0,0.0,0.0,0.0


In [28]:
asia_data = asia_data.iloc[31:]
asia_data.head()

,continent,location,date,total_cases,new_cases,total_deaths,new_deaths,total_vaccinations,people_vaccinated,people_fully_vaccinated,total_boosters,new_vaccinations
8814,NaN,Asia,22/02/20,77807.0,1719.0,2454.0,206.0,0.0,0.0,0.0,0.0,0.0
8815,NaN,Asia,23/02/20,78030.0,223.0,2463.0,9.0,0.0,0.0,0.0,0.0,0.0
8816,NaN,Asia,24/02/20,78521.0,491.0,2619.0,156.0,0.0,0.0,0.0,0.0,0.0
8817,NaN,Asia,25/02/20,79264.0,743.0,2696.0,77.0,0.0,0.0,0.0,0.0,0.0
8818,NaN,Asia,26/02/20,80063.0,799.0,2753.0,57.0,0.0,0.0,0.0,0.0,0.0


In [29]:
north_america_data = north_america_data.iloc[31:]
north_america_data.head()

,continent,location,date,total_cases,new_cases,total_deaths,new_deaths,total_vaccinations,people_vaccinated,people_fully_vaccinated,total_boosters,new_vaccinations
121567,NaN,North America,22/02/20,30.0,1.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
121568,NaN,North America,23/02/20,31.0,1.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
121569,NaN,North America,24/02/20,32.0,1.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
121570,NaN,North America,25/02/20,35.0,3.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
121571,NaN,North America,26/02/20,36.0,1.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0


In [30]:
africa_data = africa_data.iloc[9:]
africa_data.head()

,continent,location,date,total_cases,new_cases,total_deaths,new_deaths,total_vaccinations,people_vaccinated,people_fully_vaccinated,total_boosters,new_vaccinations
809,NaN,Africa,22/02/20,1.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
810,NaN,Africa,23/02/20,1.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
811,NaN,Africa,24/02/20,1.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
812,NaN,Africa,25/02/20,2.0,1.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
813,NaN,Africa,26/02/20,2.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0


In [31]:
europe_data = europe_data.iloc[30:]
europe_data.head()

,continent,location,date,total_cases,new_cases,total_deaths,new_deaths,total_vaccinations,people_vaccinated,people_fully_vaccinated,total_boosters,new_vaccinations
53364,NaN,Europe,22/02/20,121.0,42.0,3.0,1.0,0.0,0.0,0.0,0.0,0.0
53365,NaN,Europe,23/02/20,220.0,99.0,4.0,1.0,0.0,0.0,0.0,0.0,0.0
53366,NaN,Europe,24/02/20,296.0,76.0,8.0,4.0,0.0,0.0,0.0,0.0,0.0
53367,NaN,Europe,25/02/20,403.0,107.0,11.0,3.0,0.0,0.0,0.0,0.0,0.0
53368,NaN,Europe,26/02/20,561.0,158.0,14.0,3.0,0.0,0.0,0.0,0.0,0.0


In [32]:
oceania_data = oceania_data.iloc[28:]
oceania_data.head()

,continent,location,date,total_cases,new_cases,total_deaths,new_deaths,total_vaccinations,people_vaccinated,people_fully_vaccinated,total_boosters,new_vaccinations
124921,NaN,Oceania,22/02/20,15.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
124922,NaN,Oceania,23/02/20,15.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
124923,NaN,Oceania,24/02/20,15.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
124924,NaN,Oceania,25/02/20,15.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
124925,NaN,Oceania,26/02/20,15.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0


# **GOOD WORK** Date is sarting from 22/02/22

In [33]:
# checkhing lengths Again
print(f'''

        World         => {world_data['location'].count()}
        Asia          => {asia_data['location'].count()}
        Europe        => {europe_data['location'].count()}
        Africa        => {africa_data['location'].count()}
        Oceania       => {oceania_data['location'].count()}
        North America => {north_america_data['location'].count()}
        South America => {south_america_data['location'].count()}

       ''')
        



        World         => 802
        Asia          => 802
        Europe        => 802
        Africa        => 802
        Oceania       => 802
        North America => 802
        South America => 802

       


### Joining continents

In [35]:
continent_data = pd.concat((world_data, asia_data, europe_data, africa_data, south_america_data, north_america_data, oceania_data))
continent_data.sample()

,continent,location,date,total_cases,new_cases,total_deaths,new_deaths,total_vaccinations,people_vaccinated,people_fully_vaccinated,total_boosters,new_vaccinations
153160,NaN,South America,06/08/21,35856763.0,68365.0,1099855.0,1542.0,288359049.0,201270651.0,94871598.0,4291.0,3208029.0


In [42]:
# Removing continent column
continent_data.drop(['continent'], axis=1, inplace=True)
continent_data.head()

,location,date,total_cases,new_cases,total_deaths,new_deaths,total_vaccinations,people_vaccinated,people_fully_vaccinated,total_boosters,new_vaccinations
180475,World,22/02/20,78608.0,1762.0,2459.0,207.0,0.0,0.0,0.0,0.0,0.0
180476,World,23/02/20,78990.0,382.0,2470.0,11.0,0.0,0.0,0.0,0.0,0.0
180477,World,24/02/20,79558.0,568.0,2630.0,160.0,0.0,0.0,0.0,0.0,0.0
180478,World,25/02/20,80412.0,854.0,2710.0,80.0,0.0,0.0,0.0,0.0,0.0
180479,World,26/02/20,81385.0,973.0,2771.0,61.0,0.0,0.0,0.0,0.0,0.0


### Now removing continet and world data from original data

In [45]:
data.isnull().sum()

continent                  5774
location                      0
date                          0
total_cases                   0
new_cases                     0
total_deaths                  0
new_deaths                    0
total_vaccinations            0
people_vaccinated             0
people_fully_vaccinated       0
total_boosters                0
new_vaccinations              0
dtype: int64

In [47]:
data = data.dropna()
data.isnull().sum()

continent                  0
location                   0
date                       0
total_cases                0
new_cases                  0
total_deaths               0
new_deaths                 0
total_vaccinations         0
people_vaccinated          0
people_fully_vaccinated    0
total_boosters             0
new_vaccinations           0
dtype: int64

### Now saving both data sets seprately

In [50]:
continent_data.to_csv('continent_covid.csv')
data.to_csv('country_covid.csv')